### ISSUE TITLE
New rule: no-this-in-route-templates

### ISSUE THREAD

**Issue by jelhan** [ember-template-lint/ember-template-lint#1585]
I like to propose a new rule: `no-this-in-route-templates`. It would forbid using `this` context explicitly in route templates.

## Motivation

The rule enforces route templates to be treated as template-only components. It would allow teams to opt out of explicit controller usage.

Eslint-plugin-ember has a [`@ember/no-controller` rule](https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/no-controllers.md), which shares the same motivation. It disallow explicit definition of a controller unless it's required for query params.

Even if `@ember/no-controller` ESlint rule is used state could be read and modified from the route template. This could undermine the motivation of the rule as shown in this example:

```js
import Controller from '@ember/controller';

export default FooController extends Controller {
  queryParams = ['sortBy'];

  sortBy = 'name';
}
```

```hbs
Currently sorted by {{this.sortBy}}.
<button {{on "click" (action (mut this.sortBy) "city")}}>Sort by city insteady</button>

Current name: {{this.name}}
<input value={{this.name}} {{on "input" (action (mut this.name) value="target.value"}} />
```

The controller would be valid in regards of `@ember/no-controller` rule. But the template implementation would still highly rely on controllers.

The proposed `no-this-in-route-templates` rule would prevent that as long as used together with the [`no-implicit-this` rule](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-implicit-this.md).

## Detailed design

The rule would only affect route templates. It would ignore all component templates. A helper to distinguish route and component templates exists already: https://github.com/ember-template-lint/ember-template-lint/blob/master/lib/helpers/is-route-template.js

The rule should be an optional rule for the foreseeable future.

It should only lint against explicit `{{this}}` usage to reduce complexity. Consumers should be encouraged to use it together with `no-implicit-this` rule.

## Examples

**Good**:

```hbs
{{! app/templates/foo.hbs}}

{{@model}}
```

```hbs
{{! app/components/foo.hbs }}

{{this.model}}
```

```hbs
{{! app/template/components/foo.hbs}}

{{this.model}}
```

**Bad**:

```hbs
{{! app/templates/foo.hbs}}

{{this.model}}
```

**Comment by rwjblue** [comment]
Seems good to me, though the exact mechanism that we use to determine if a template is a route template or a component template is a bit silly (just a regexp against the template file path).

**Comment by TSenter** [comment]
Would this conflict with the [`no-model-argument-in-route-templates`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-model-argument-in-route-templates.md) rule?
